src/lib/datepicker/datepicker.ts

```diff
@@ -385,73 +385,84 @@ export class MatDatepicker<D> implements OnDestroy, CanColor {
     }
   }
 
-  /** Open the calendar as a dialog. */
-  private _openAsDialog(): void {
-    // Usually this would be handled by `open` which ensures that we can only have one overlay
-    // open at a time, however since we reset the variables in async handlers some overlays
-    // may slip through if the user opens and closes multiple times in quick succession (e.g.
-    // by holding down the enter key).
-    if (this._dialogRef) {
-      this._dialogRef.close();
+  /** Opens the overlay with the calendar. */
+  private _openOverlay(): void {
+    const overlayRef = this.touchUi ? this._getDialogRef() : this._getPopupRef();
+
+    // Don't do anything if the relevant overlay is already attached.
+    if (overlayRef.hasAttached()) {
+      return;
     }
 
-    this._dialogRef = this._dialog.open<MatDatepickerContent<D>>(MatDatepickerContent, {
-      direction: this._dir ? this._dir.value : 'ltr',
-      viewContainerRef: this._viewContainerRef,
-      panelClass: 'mat-datepicker-dialog',
-    });
+    this._componentRef = overlayRef.attach(this._getCalendarPortal());
+    this._componentRef.instance.datepicker = this;
+    this._componentRef.instance.color = this.color;
+
+    if (!this.touchUi) {
+      // Update the position once the calendar has rendered.
+      this._ngZone.onStable.asObservable()
```

Comment:
I think we can leave off the `.asObservable()`

Reply:
I've been doing this since, at some point in the past, omitting it ended up causing a memory leak. I'm not sure whether that hasn't been fixed in core by now though.

Reply:
Ah ok, didn't realize that